### ISSUE TITLE
Add metadata schema validation

### ISSUE THREAD

**Issue by lukpueh** [in-toto/in-toto-golang#21]
**Description of issue or feature request**:
`model.go` defines structs that may be used to hold in-toto metadata. The metadata schema is only loosely prescribed by the used datatypes and by the specified json tags.

We need a generic schema validation facility that allows to verify if a given instance of a struct is conformant with the [in-toto metadata specification](https://github.com/in-toto/docs/blob/v0.9/in-toto-spec.md#4-document-formats). 

For inspiration take a look at the reference implementation, where we use a [`ValidationMixin`](https://github.com/in-toto/in-toto/blob/v0.2.3/in_toto/models/common.py#L27-L54) for top-level metadata models, plus [`securesystemslib.schema`](https://github.com/secure-systems-lab/securesystemslib/blob/v0.11.2/securesystemslib/schema.py)-checkers to validate mostly (but not exclusively) crypto-related data structures, see as defined in ([`securesystemslib.formats`](https://github.com/secure-systems-lab/securesystemslib/blob/v0.11.2/securesystemslib/formats.py), [`in_toto.formats`](https://github.com/in-toto/in-toto/blob/v0.2.3/in_toto/formats.py) and [`in_toto.gpg.formats`](https://github.com/in-toto/in-toto/blob/v0.2.3/in_toto/gpg/formats.py)).

**Current behavior**:
No methods to validate the data model.

**Expected behavior**:
Provide methods to validate the data model and use where appropriate (e.g. to check function parameters).

**Comment by adityasaky** [comment]
I'm working on this.

**Comment by adityasaky** [comment]
I was looking at the validations being carried out in the reference implementation, specifically in `Layout` and `Link`.

- In both, there's a check for the value of `type` is what it should be - this is a valid check.
- In `Layout`, the expiry date is validated - both the format and if it's a an actual valid date. Again, this is a valid check.
- In `Layout`, readme is checked to ensure it's a string - this is redundant in golang as far as I can tell.
- In `Layout`, the keys are checked mostly for the format of the keyid and the key itself. 
- In `Layout`, the steps and inspections are checked. The checks are for type and uniqueness - again, checking for the type feels redundant in golang. Checking uniqueness is definitely important.
- In `Link`, the materials, products, and byproducts are checked for their type, and are independently validated. The type check again feels redundant, however, I'm going to look at the validate calls for steps and inspections and see what's happening there next.

Thanks to @lukpueh and @SantiagoTorres for pointing me in the right direction regarding json.Unmarshal too.

**Comment by lukpueh** [comment]
@adityasaky, this sounds good so far, please also take a look at the crypto-related format checkers (private and public only! keys, signatures, signature schemas, keyids, ...).

Moreover, it might be a good idea to check if metadata loaded from json has any additional non-conformant entries. Currently (here and I think in python too) we just silently ignore them. we might want to warn at least.

**Comment by adityasaky** [comment]
Could you expand on the non-conformant entries? Do you mean incorrect values? The types are all easily handled and golang itself will error out.

Key validations seem to be the ~key~ primary checks, apart from ensuring step and inspection names remain unique.

**Comment by lukpueh** [comment]
I was referring to additional incorrect fields. Take an in-toto metadata file like:
```
{
  "signatures": [<list of validly formatted signature objects>],
  "signed": <validly formatted link or layout object>,
  "foo": "bar"
}
```
Here and in python our metadata parser would just silently ignore the additional field "foo". 

As a matter of fact, @awwad has pointed out this issue a while ago and we even have an open ticket for it: https://github.com/in-toto/in-toto/issues/186

**Comment by lukpueh** [comment]
Fixed with #25. Thanks for your hard work, @adityasaky!
